phabricatoremails/render/events/common.py

```diff
@@ -0,0 +1,79 @@
+from dataclasses import dataclass
+from datetime import timezone, timedelta
+from enum import Enum
+from math import floor, ceil
+from typing import Optional, Dict, List
+
+"""Describes data structures that are used by both secure and public revision events.
+
+Additional information about this module can be found in __init__.py
+"""
+
+
+@dataclass
+class Recipient:
+    email: str
+    timezone: timezone
+    is_actor: bool
+
+    @staticmethod
+    def parse(recipient: Dict):
```

Comment:
Hmm, I'm not sure that this is an improvement for a few reasons:

1. Though explicitly grabbing properties may seem menial (`return Reviewer(reviewer["name"], reviewer["isActionable"], ...)`), it's _simple_. It's easy to see what's happening.

The `key_mapping` technique reduces some raw keystrokes, but it's difficult to understand what's happening at a glance, e.g.: "What do the keys and values of that map do? I don't specify properties on my class, but they're somehow set? Why isn't my IDE/tooling able to tell me about the fields of my class?".

This comes back to the "using code during read-time vs write-time" [discussion from earlier](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r405789390).

2. The `parse()` function currently handles mapping from transfer data types to "usable" data types. For example, `timezone_offset_seconds` isn't as useful to us as a `timezone`. Once we've parsed to the new value, then having the old one around isn't valuable. If we move all of the conversion logic to `@property` functions, then our objects are going to have extra state on them that isn't valuable:

![extra-state](https://user-images.githubusercontent.com/7784737/79160244-4f733600-7d8e-11ea-8e14-b0a9a6845bea.png)

Reply:
> Though explicitly grabbing properties may seem menial (return Reviewer(reviewer["name"], reviewer["isActionable"], ...)), it's simple. It's easy to see what's happening.

It's simpler to see what a "base" `Event.parse` is doing and understand that all events are doing the same thing, rather than having to go through each event individually. You can just learn how to use an `Event` class and apply it to all events.

> The parse() function currently handles mapping from transfer data types to "usable" data types. For example, timezone_offset_seconds isn't as useful to us as a timezone. Once we've parsed to the new value, then having the old one around isn't valuable. If we move all of the conversion logic to @property functions, then our objects are going to have extra state on them that isn't valuable:

In that case, you can probably establish a naming convention (e.g. `_timezone_offset_seconds`). You never know when someone wants to do some debugging and is looking for a particular value.

Moving these manually calculated values to properties also makes it clearer as to what is a straight-through translation, and what is calculated based on other fields.

Reply:
While writing tests for some of the templates, I'm realizing another benefit of parsing up front rather than having `@property` methods: construction of mock events/revisions/etc is easier. For example, I can provide the recipient's timezone with a relevant data type (`timezone.utc`) rather than primitive values (`0`).